File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.java
class Solution {
    public List<String> letterCombinations(String digits) {
        List<String> ll = new ArrayList<>();
        if(digits.length()==0)
        {
            return ll;
        }
        printsol(code,digits,"",ll);
        return ll;
    }

    static String code[] = {"","","abc","def","ghi","jkl","mno","pqrs","tuv","wxyz"};

    public static void printsol(String code[],String ques,String ans,List<String> ll)
	{
		if(ques.length()==0)
		{
			ll.add(ans);
			return;
		}
		char ch = ques.charAt(0);
		String press = code[ch-48];
		for(int i=0;i<press.length();i++)
		{
			printsol(code,ques.substring(1),ans+press.charAt(i),ll);
		}
	}
}

File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        self.result = []
        
        if len(digits) == 0:
            return self.results
        
        num_map = ["","abc","def","ghi","jkl","mno","pqrs","tuv","wxyz"]

        self.solve(digits,num_map,[],0)
        return self.result
    
    def solve(self,digits,num_map,subans,idx):
        if idx >= len(digits):
            self.result.append("".join(subans))
            return

        for ch in num_map[int(digits[idx])-1]:
            subans.append(ch)
            self.solve(digits,num_map,subans,idx+1)
            subans.pop()
        